File: P2/C1/lista.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

// Definición del nodo
typedef struct Nodo {
    char* dato;              // Cadena dinámica
    struct Nodo* siguiente;
} Nodo;

// Función para agregar al final
void agregar(Nodo** cabeza, const char* texto) {
    // Crear nuevo nodo
    Nodo* nuevo_nodo = (Nodo*) malloc(sizeof(Nodo));
    if (nuevo_nodo == NULL) {
        printf("Error al asignar memoria para el nodo.\n");
        return;
    }

    // Reservar memoria exacta para la cadena
    nuevo_nodo->dato = (char*) malloc(strlen(texto) + 1);
    if (nuevo_nodo->dato == NULL) {
        printf("Error al asignar memoria para la cadena.\n");
        free(nuevo_nodo);
        return;
    }

    strcpy(nuevo_nodo->dato, texto);
    nuevo_nodo->siguiente = NULL;

    // Si la lista está vacía
    if (*cabeza == NULL) {
        *cabeza = nuevo_nodo;
        return;
    }

    // Recorrer hasta el final
    Nodo* actual = *cabeza;
    while (actual->siguiente != NULL) {
        actual = actual->siguiente;
    }

    actual->siguiente = nuevo_nodo;
}

// Mostrar lista
void mostrar(Nodo* cabeza) {
    Nodo* actual = cabeza;

    while (actual != NULL) {
        printf("%s\n", actual->dato);
        actual = actual->siguiente;
    }
}

// Liberar lista completamente
void liberarLista(Nodo* cabeza) {
    Nodo* temp;

    while (cabeza != NULL) {
        temp = cabeza;
        cabeza = cabeza->siguiente;

        free(temp->dato);   // Liberar la cadena primero
        free(temp);         // Luego el nodo
    }
}

int main() {
    Nodo* cabeza = NULL;

    agregar(&cabeza, "Comprar pan");
    agregar(&cabeza, "Enviar correo");
    agregar(&cabeza, "Estudiar listas enlazadas en C");

    printf("Lista de tareas:\n");
    mostrar(cabeza);

    liberarLista(cabeza);

    return 0;
}

File: P2/C1/lista.py
# Ejemplo de lista enlazada simple para una lista de tareas

class Nodo:
    def __init__(self, dato):
        self.dato = dato      # Información que guarda el nodo
        self.siguiente = None # Puntero al siguiente nodo

class ListaEnlazada:
    def __init__(self):
        self.cabeza = None  # Primer nodo de la lista

    def agregar(self, dato):
        """Agrega un nuevo nodo al final de la lista."""
        nuevo_nodo = Nodo(dato)
        if not self.cabeza:
            self.cabeza = nuevo_nodo
            return
        actual = self.cabeza
        while actual.siguiente:
            actual = actual.siguiente
        actual.siguiente = nuevo_nodo

    def mostrar(self):
        """Muestra todos los elementos de la lista."""
        actual = self.cabeza
        while actual:
            print(actual.dato)
            actual = actual.siguiente

# Ejemplo de uso
tareas = ListaEnlazada()
tareas.agregar("Comprar pan")
tareas.agregar("Enviar correo")
tareas.agregar("Estudiar listas enlazadas")

print("Lista de tareas:")
tareas.mostrar()